climpred/stats.py

```diff
@@ -155,71 +155,63 @@ def xr_rm_poly(ds, order, dim='time'):
-        # affect the original dataset that is entered.
-        dims = ds.copy().dims
-        coords = ds.copy().coords
-        return dims, coords
-
-    def _swap_axes(y):
-        """
-        Push the independent axis up to the first dimension if needed. E.g.,
-        the user submits a DataArray with dimensions ('lat','lon','time'), but
-        wants the regression performed over time. This function expects the
-        leading dimension to be the independent axis, so this subfunction just
-        moves it up front.
-        """
-        dims, coords = _get_metadata(y)
-        if dims[0] != dim:
-            idx = dims.index(dim)
-            y = np.swapaxes(y, 0, idx)
-            # fixes bug with renaming dimensions that have
-            # coordinates with differing shapes.
-            del y.coords[dim], y.coords[dims[0]]
-            y = y.rename({dims[0]: dim,
-                          dims[idx]: dims[0]})
-            dims = list(dims)
-            dims[0], dims[idx] = dims[idx], dims[0]
-            dims = tuple(dims)
-        return np.asarray(y), dims, coords
-
-    def _reconstruct_ds(y, store_vars):
-        """
-        Rebuild the original dataset.
-        """
-        new_ds = xr.Dataset()
-        for i, var in enumerate(store_vars):
-            new_ds[var] = xr.DataArray(y.isel(variable=i))
-        return new_ds
+    if dim not in ds.dims:
+        raise KeyError(
+            f"Input dim, '{dim}', was not found in the ds; "
+            f"found only the following dims: {list(ds.dims)}."
+        )
 
+    # handle both datasets and dataarray
     if isinstance(ds, xr.Dataset):
-        DATASET, store_vars = True, ds.data_vars
-        y = []
-        for var in ds.data_vars:
-            y.append(ds[var])
-        y = xr.concat(y, 'variable')
+        da = ds.to_array()
+        return_ds = True
+    else:
+        da = ds.copy()
+        return_ds = False
+
+    da_dims_orig = list(da.dims)  # orig -> original
+    if len(da_dims_orig) > 1:
+        # want independent axis to be the leading dimension
+        da_dims_swap = da_dims_orig.copy()  # copy to prevent contamination
+
+        # https://stackoverflow.com/questions/1014523/
+        # simple-syntax-for-bringing-a-list-element-to-the-front-in-python
+        da_dims_swap.insert(0, da_dims_swap.pop(da_dims_swap.index(dim)))
+        da = da.transpose(*da_dims_swap)
+
+        # hide other dims into a single dim
+        da = da.stack({'other_dims': da_dims_swap[1:]})
+        dims_swapped = True
     else:
-        DATASET = False
-        y = ds
-    # Force independent axis to be leading dimension.
-    y, dims, coords = _swap_axes(y)
-    data_shape = y.shape
-    y = y.reshape((data_shape[0], -1))
-    # NaNs screw up vectorized regression; just fill with zeros.
-    y[np.isnan(y)] = 0
+        dims_swapped = False
+
+    # NaNs will make the polyfit fail; fill with 0s (will 0 affect the fit?)
```

Comment:
The assumption here is that if a NaN is encountered, it is a masked area. For instance, a NaN would represent a land grid cell in ocean model output. So the whole time series at this grid cells is NaNs, therefore we don't need a fit there and will remask in the end. However, I'm not sure how this would affect a time series with NaNs within it. That would definitely affect the fit, although maintaining a NaN in the time series would break the polyfit?

Reply:
the nans are aweful. I havent figured out a way how to easily not work on grid cells with once have nan and not break the code.